File: src/main/scala/mffs/tile/TileFortronCapacitor.java
package mffs.tile;

import resonant.api.mffs.card.ICard;
import resonant.api.mffs.card.ICardInfinite;
import resonant.api.mffs.card.ICoordLink;
import resonant.api.mffs.fortron.FrequencyGrid;
import resonant.api.mffs.fortron.IFortronCapacitor;
import resonant.api.mffs.fortron.IFortronFrequency;
import resonant.api.mffs.fortron.IFortronStorage;
import resonant.api.mffs.modules.IModule;
import com.google.common.io.ByteArrayDataInput;
import mffs.MFFSHelper;
import mffs.ModularForceFieldSystem;
import mffs.TransferMode;
import mffs.base.TileModuleAcceptor;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import resonant.api.blocks.IBlockFrequency;
import universalelectricity.core.transform.vector.Vector3;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class TileFortronCapacitor extends TileModuleAcceptor implements IFortronStorage, IFortronCapacitor
{
	private TransferMode transferMode = TransferMode.EQUALIZE;

	public TileFortronCapacitor()
	{
		this.capacityBase = 700;
		this.capacityBoost = 10;
		this.startModuleIndex = 2;
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		this.consumeCost();

		/**
		 * Transmit Fortrons in frequency network, evenly distributing them.
		 */
		/**
		 * Gets the card.
		 */
		if (this.isActive() && this.ticks % 10 == 0)
		{
			Set<IFortronFrequency> machines = new HashSet<IFortronFrequency>();

			for (ItemStack itemStack : this.getCards())
			{
				if (itemStack != null)
				{
					if (itemStack.getItem() instanceof ICardInfinite)
					{
						this.setFortronEnergy(this.getFortronCapacity());
					}
					else if (itemStack.getItem() instanceof ICoordLink)
					{
						Vector3 linkPosition = ((ICoordLink) itemStack.getItem()).getLink(itemStack);

						if (linkPosition != null && linkPosition.getTileEntity(this.worldObj) instanceof IFortronFrequency)
						{
							machines.add(this);
							machines.add((IFortronFrequency) linkPosition.getTileEntity(this.worldObj));
						}
					}
				}
			}

			if (machines.size() < 1)
			{
				machines = this.getLinkedDevices();
			}

			MFFSHelper.transferFortron(this, machines, this.transferMode, getTransmissionRate() * 10);
		}
	}

	@Override
	public float getAmplifier()
	{
		return 0.001f;
	}

	/**
	 * Packet Methods
	 */
	@Override
	public ArrayList getPacketData(int packetID)
	{
		ArrayList data = super.getPacketData(packetID);

		if (packetID == TilePacketType.DESCRIPTION.ordinal())
		{
			data.add(this.transferMode.ordinal());
		}

		return data;
	}

	@Override
	public void onReceivePacket(int packetID, ByteArrayDataInput dataStream) throws IOException
	{
		super.onReceivePacket(packetID, dataStream);

		if (packetID == TilePacketType.DESCRIPTION.ordinal())
		{
			this.transferMode = TransferMode.values()[dataStream.readInt()];
		}
		else if (packetID == TilePacketType.TOGGLE_MODE.ordinal())
		{
			this.transferMode = this.transferMode.toggle();
		}
	}

	@Override
	public int getSizeInventory()
	{
		return 5;
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		this.transferMode = TransferMode.values()[nbt.getInteger("transferMode")];
	}

	@Override
	public void writeToNBT(NBTTagCompound nbttagcompound)
	{
		super.writeToNBT(nbttagcompound);
		nbttagcompound.setInteger("transferMode", this.transferMode.ordinal());
	}

	@Override
	public Set<IFortronFrequency> getLinkedDevices()
	{
		Set<IFortronFrequency> fortronBlocks = new HashSet<IFortronFrequency>();
		Set<IBlockFrequency> frequencyBlocks = FrequencyGrid.instance().get(this.worldObj, new Vector3(this), this.getTransmissionRange(), this.getFrequency());

		for (IBlockFrequency frequencyBlock : frequencyBlocks)
		{
			if (frequencyBlock instanceof IFortronFrequency)
			{
				fortronBlocks.add((IFortronFrequency) frequencyBlock);
			}
		}

		return fortronBlocks;
	}

	@Override
	public boolean isItemValidForSlot(int slotID, ItemStack itemStack)
	{
		if (slotID == 0 || slotID == 1)
		{
			return itemStack.getItem() instanceof ICard;
		}
		else
		{
			return itemStack.getItem() instanceof IModule;
		}
	}

	@Override
	public Set<ItemStack> getCards()
	{
		Set<ItemStack> cards = new HashSet<ItemStack>();
		cards.add(super.getCard());
		cards.add(this.getStackInSlot(1));
		return cards;
	}

	public TransferMode getTransferMode()
	{
		return this.transferMode;
	}

	@Override
	public int getTransmissionRange()
	{
		return 15 + this.getModuleCount(ModularForceFieldSystem.itemModuleScale);
	}

	@Override
	public int getTransmissionRate()
	{
		return 250 + 50 * this.getModuleCount(ModularForceFieldSystem.itemModuleSpeed);
	}
}

File: src/main/scala/mffs/production/TileFortronCapacitor.scala
package mffs.production

import java.util.{HashSet => JHashSet, Set => JSet}

import cpw.mods.fml.relauncher.{Side, SideOnly}
import io.netty.buffer.ByteBuf
import mffs.Content
import mffs.base.{TileModuleAcceptor, TilePacketType}
import mffs.item.card.ItemCardFrequency
import mffs.util.TransferMode.TransferMode
import mffs.util.{FortronUtility, TransferMode}
import net.minecraft.client.renderer.RenderBlocks
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.fluids.IFluidContainerItem
import resonant.api.mffs.card.ICoordLink
import resonant.api.mffs.fortron.{FrequencyGridRegistry, IFortronCapacitor, IFortronFrequency, IFortronStorage}
import resonant.api.mffs.modules.IModule
import resonant.lib.network.ByteBufWrapper.ByteBufWrapper
import resonant.lib.network.discriminator.PacketType
import universalelectricity.core.transform.vector.Vector3

import scala.collection.JavaConversions._

class TileFortronCapacitor extends TileModuleAcceptor with IFortronStorage with IFortronCapacitor
{
  private var transferMode = TransferMode.equalize
  private val tickRate = 10

  capacityBase = 700
  capacityBoost = 10
  startModuleIndex = 1

  override def getSizeInventory = 3 + 4 * 2 + 1

  override def update()
  {
    super.update()
    this.consumeCost()

    if (isActive)
    {
      /**
       * Draw from input slots and eject from output slots
       */
      getInputStacks filter (_.getItem.isInstanceOf[IFluidContainerItem]) foreach (stack => fortronTank.fill(stack.getItem.asInstanceOf[IFluidContainerItem].drain(stack, Math.min(getFortronEmpty, getTransmissionRate), true), true))

      if (fortronTank.getFluidAmount > 0)
      {
        val transferFluid = fortronTank.getFluid.copy()
        transferFluid.amount = Math.min(transferFluid.amount, getTransmissionRate)
        getOutputStacks filter (_.getItem.isInstanceOf[IFluidContainerItem]) foreach (stack => fortronTank.drain(stack.getItem.asInstanceOf[IFluidContainerItem].fill(stack, transferFluid, true), true))
      }

      if (ticks % tickRate == 0)
      {
        /**
         * Transfer based on input/output slots
         */
        FortronUtility.transferFortron(this, getInputDevices, TransferMode.fill, getTransmissionRate * tickRate)
        FortronUtility.transferFortron(this, getOutputDevices, TransferMode.drain, getTransmissionRate * tickRate)

        /**
         * Transfer based on frequency
         */
        FortronUtility.transferFortron(this, getFrequencyDevices, transferMode, getTransmissionRate * tickRate)
      }
    }
  }

  override def getAmplifier: Float = 0f

  /**
   * Packet Methods
   */

  override def write(buf: ByteBuf, id: Int)
  {
    super.write(buf, id)

    if (id == TilePacketType.description.id)
    {
      buf <<< transferMode.id
    }
  }

  override def read(buf: ByteBuf, id: Int, player: EntityPlayer, packet: PacketType) : Boolean =
  {
	  super.read(buf, id, player, packet)

    if (id == TilePacketType.description.id)
    {
      transferMode = TransferMode(buf.readInt)
    }
    else if (id == TilePacketType.toggleMoe.id)
    {
      transferMode = this.transferMode.toggle
    }

    return false
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    this.transferMode = TransferMode(nbt.getInteger("transferMode"))
  }

  override def writeToNBT(nbttagcompound: NBTTagCompound)
  {
    super.writeToNBT(nbttagcompound)
    nbttagcompound.setInteger("transferMode", this.transferMode.id)
  }

  def getDeviceCount = getFrequencyDevices.size + getInputDevices.size + getOutputDevices.size

  override def getFrequencyDevices: JSet[IFortronFrequency] = FrequencyGridRegistry.instance.getNodes(classOf[IFortronFrequency], world, new Vector3(this), getTransmissionRange, getFrequency)

  def getInputDevices: JSet[IFortronFrequency] = getDevicesFromStacks(getInputStacks)

  def getOutputDevices: JSet[IFortronFrequency] = getDevicesFromStacks(getOutputStacks)

  def getDevicesFromStacks(stacks: Set[ItemStack]): JSet[IFortronFrequency] =
  {
    val devices = new JHashSet[IFortronFrequency]()

    stacks
            .filter(_.getItem.isInstanceOf[ICoordLink])
            .map(itemStack => itemStack.getItem.asInstanceOf[ICoordLink].getLink(itemStack))
            .filter(linkPosition => linkPosition != null && linkPosition.getTileEntity(world).isInstanceOf[IFortronFrequency])
            .foreach(linkPosition => devices.add(linkPosition.getTileEntity(world).asInstanceOf[IFortronFrequency]))

    return devices
  }

  override def isItemValidForSlot(slotID: Int, itemStack: ItemStack): Boolean =
  {
    if (slotID == 0)
    {
      return itemStack.getItem.isInstanceOf[ItemCardFrequency]
    }
    else if (slotID < 4)
    {
      return itemStack.getItem.isInstanceOf[IModule]
    }

    return true
  }

  def getInputStacks: Set[ItemStack] = ((4 to 7) map (i => getStackInSlot(i)) filter (_ != null)).toSet

  def getOutputStacks: Set[ItemStack] = ((8 to 11) map (i => getStackInSlot(i)) filter (_ != null)).toSet

  def getTransferMode: TransferMode = transferMode

  def getTransmissionRange: Int = 15 + getModuleCount(Content.moduleScale)

  def getTransmissionRate: Int = 300 + 60 * getModuleCount(Content.moduleSpeed)

  @SideOnly(Side.CLIENT)
  override def renderStatic(renderer: RenderBlocks, pos: Vector3, pass: Int): Boolean =
  {
    return false
  }

  @SideOnly(Side.CLIENT)
  override def renderDynamic(pos: Vector3, frame: Float, pass: Int)
  {
    RenderFortronCapacitor.render(this, pos.x, pos.y, pos.z, frame, isActive, false)
  }

  @SideOnly(Side.CLIENT)
  override def renderInventory(itemStack: ItemStack)
  {
    RenderFortronCapacitor.render(this, -0.5, -0.5, -0.5, 0, true, true)
  }
}